Source/Core/InputCommon/ControlReference/FunctionExpression.cpp

```diff
@@ -636,6 +636,57 @@ class PulseExpression : public FunctionExpression
   mutable Clock::time_point m_release_time = Clock::now();
 };
 
+// usage: ntimes(input, n = 1)
+class NTimesExpression : public FunctionExpression
+{
+private:
+  ArgumentValidation
+  ValidateArguments(const std::vector<std::unique_ptr<Expression>>& args) override
+  {
+    if (args.size() == 1 || args.size() == 2)
+      return ArgumentsAreValid{};
+    else
+      return ExpectedArguments{"input, times = 1"};
+  }
+
+  ControlState GetValue() const override
+  {
+    if (m_running)
+    {
+      m_loops++;
+      if (m_loops >= m_times * 2 - 1)
+        m_running = false;
+
+      if (m_loops % 2 == 1)
+        return m_input;
+      else
+        return 0.0;
+    }
+    else if (m_zeroed && GetArg(0).GetValue() != 0.0)
+    {
+      m_input = GetArg(0).GetValue();
+      m_times = GetArg(1).GetValue();
```

Comment:
For consistency, all function expressions should read and cache all their possible arguments at the top.
This is in case the arg is also another expression which modifies its state in its GetValue(), so it needs to be called for every call of this function GetValue(), not only when some conditions are true.

Reply:
Thanks, I think I see what you mean. I've inserted new `const ControlState`s `input` and `times` at the top of `GetValue()` to store the current inputs while retaining the original inputs in `m_input` and `m_times` - does that cover it?